core/vmpu/src/armv7m/vmpu_armv7m.c

```diff
@@ -127,29 +127,45 @@ void vmpu_sys_mux_handler(uint32_t lr, uint32_t msp)
 
     switch (ipsr) {
         case MemoryManagement_IRQn:
-            /* Currently we only support recovery from unprivileged mode. */
-            if (lr & 0x4) {
+            fault_status = VMPU_SCB_MMFSR;
+
+            /* If we are having an unstacking fault, we can't read the pc
+             * at fault. */
+            if (fault_status & 0x18) {
+                /* fake pc */
+                pc = 0x0;
+
+                /* The stack pointer is at fault. MMFAR doesn't contain a
+                 * valid fault address. */
+                fault_addr = lr & 0x4 ? psp : msp;
+            } else {
                 /* pc at fault */
-                pc = vmpu_unpriv_uint32_read(psp + (6 * 4));
+                if (lr & 0x4) {
+                    pc = vmpu_unpriv_uint32_read(psp + (6 * 4));
+                } else {
+                    /* We can be privileged here if we tried doing an ldrt or
+                     * strt to a region not currently loaded in the MPU. In
+                     * such cases, we are reading from the msp and shouldn't go
+                     * through vmpu_unpriv_uint32_read. A box wouldn't have
+                     * access to our stack. */
+                    pc = *(uint32_t *) (msp + (6 * 4));
+                }
 
                 /* Backup fault address and status */
                 fault_addr = SCB->MMFAR;
-                fault_status = VMPU_SCB_MMFSR;
-
-                /* Check if the fault is an MPU fault. */
-                if (vmpu_fault_recovery_mpu(pc, psp, fault_addr, fault_status)) {
-                    VMPU_SCB_MMFSR = fault_status;
-                    return;
-                }
+            }
 
-                /* If recovery was not successful, throw an error and halt. */
-                DEBUG_FAULT(FAULT_MEMMANAGE, lr, psp);
+            /* Check if the fault is an MPU fault. */
+            if (vmpu_fault_recovery_mpu(pc, psp, fault_addr, fault_status)) {
                 VMPU_SCB_MMFSR = fault_status;
-                HALT_ERROR(PERMISSION_DENIED, "Access to restricted resource denied");
-            } else {
-                DEBUG_FAULT(FAULT_MEMMANAGE, lr, msp);
-                HALT_ERROR(FAULT_MEMMANAGE, "Cannot recover from privileged MemManage fault");
+                return;
             }
+
+            /* If recovery was not successful, throw an error and halt. */
+            DEBUG_FAULT(FAULT_MEMMANAGE, lr, lr & 0x4 ? psp : msp);
+            VMPU_SCB_MMFSR = fault_status;
+            HALT_ERROR(PERMISSION_DENIED, "Access to restricted resource denied");
+            debug_box_enter_from_priv(ipsr, lr);
```

Comment:
I didn't like it either. I justified it to myself my thinking, "We are halting anyway. We don't care if we leave anything on the stack."

Your suggestion is an improvement in the sense that the stack is cleaned up. I can see the importance of leaving the uVisor stack clean when we support halting individual boxes.